From ee3ff60f0bf8e17835d6259a902beaa7af8c638e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Samuel=20Mar=C3=ADn=20Soto?= <samarin@estudiantec.cr>
Date: Sat, 26 Sep 2026 12:04:00 -0600
Subject: [PATCH] Only check STAT designAxisSize when design axes are present

A bad designAxisSize is harmless when designAxisCount is zero, since
no axis records follow the header; only validate it before reading
axes. Matches upstream commit 77c019af ("Don't check STAT
designAxisSize if the designAxisCount is zero") as carried in UXP's
vendored snapshot.
---
 src/stat.cc | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/src/stat.cc b/src/stat.cc
index 9b78281..4648dc6 100644
--- a/src/stat.cc
+++ b/src/stat.cc
@@ -53,10 +53,6 @@ bool OpenTypeSTAT::Parse(const uint8_t* data, size_t length) {
     this->minorVersion = 2;
   }
 
-  if (this->designAxisSize < sizeof(AxisRecord)) {
-    return Drop("Invalid designAxisSize");
-  }
-
   size_t headerEnd = table.offset();
 
   if (this->designAxisCount == 0) {
@@ -65,9 +61,13 @@ bool OpenTypeSTAT::Parse(const uint8_t* data, size_t length) {
       this->designAxesOffset = 0;
     }
   } else {
+    if (this->designAxisSize < sizeof(AxisRecord)) {
+      return Drop("Invalid designAxisSize");
+    }
     if (this->designAxesOffset < headerEnd ||
-        size_t(this->designAxesOffset) +
-          size_t(this->designAxisCount) * size_t(this->designAxisSize) > length) {
+        size_t(this->designAxesOffset) > length ||
+        size_t(this->designAxisCount) * size_t(this->designAxisSize) >
+          length - size_t(this->designAxesOffset)) {
       return Drop("Invalid designAxesOffset");
     }
   }
-- 
2.55.0

